src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -11715,6 +11735,298 @@ void Test1(object x1, object? y1)
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(13, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_NullConstant()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        const string? nullConstant = null;
+        if (x is nullConstant)
+        {
+            x.ToString(); // warn
+        }
+        else
+        {
+            x.ToString();
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (9,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_Null_AlreadyTestedAsNonNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        if (x != null)
+        {
+            if (x is null) // hidden
+            {
+                x.ToString(); // warn
+            }
+            else
+            {
+                x.ToString();
+            }
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (8,22): hidden CS8606: Result of the comparison is possibly always false.
+                //             if (x is null) // hidden
+                Diagnostic(ErrorCode.HDN_NullCheckIsProbablyAlwaysFalse, "null").WithLocation(8, 22),
+                // (10,17): warning CS8602: Possible dereference of a null reference.
+                //                 x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(10, 17)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsConstantPattern_Null_AlreadyTestedAsNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        if (x == null)
+        {
+            if (x is null)
+            {
+                x.ToString(); // warn
+            }
+            else
+            {
+                x.ToString();
+            }
+        }
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (10,17): warning CS8602: Possible dereference of a null reference.
+                //                 x.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(10, 17)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_IsDeclarationPattern()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test(object? x)
+    {
+        if (x is C c)
```

Comment:
Consider testing `x is T t` where `T` is unconstrained or constrained to `struct` or `class`.